flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -417,12 +417,50 @@ public void cancel() {
 	@Override
 	public final boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {
 		final Optional<ExecutionVertexID> executionVertexId = getExecutionVertexId(taskExecutionState.getID());
-		if (executionVertexId.isPresent()) {
-			executionGraph.updateState(taskExecutionState);
-			updateTaskExecutionStateInternal(executionVertexId.get(), taskExecutionState);
+
+		boolean updateSuccess = executionGraph.updateState(taskExecutionState);
+
+		if (updateSuccess && executionVertexId.isPresent()) {
+			final Optional<TaskExecutionState> validTaskExecutionState = getValidTaskExecutionState(
+				executionVertexId.get(),
+				taskExecutionState);
+
+			if (validTaskExecutionState.isPresent()) {
+				updateTaskExecutionStateInternal(executionVertexId.get(), validTaskExecutionState.get());
+			}
 			return true;
+		} else {
+			return false;
 		}
-		return false;
+	}
+
+	private Optional<TaskExecutionState> getValidTaskExecutionState(
+			final ExecutionVertexID executionVertexId,
+			final TaskExecutionState taskExecutionState) {
+
+		final ExecutionVertex executionVertex = getExecutionVertex(executionVertexId);
+		final ExecutionAttemptID currentExecutionAttemptID = executionVertex.getCurrentExecutionAttempt().getAttemptId();
+
+		// only notifies FINISHED and FAILED states which are needed at the moment.
+		// can be refined in FLINK-14233 after the legacy scheduler is removed and
+		// the actions are factored out from ExecutionGraph.
+		switch (taskExecutionState.getExecutionState()) {
+			case FINISHED:
+			case FAILED:
+				if (executionVertex.getExecutionState() == taskExecutionState.getExecutionState()) {
+					return Optional.of(
+						new TaskExecutionState(
```

Comment:
Do we actually need to create a new `TaskExecutionState` instance? It seems that all fields match at this point. How about something like:

```
	@Override
	public final boolean updateTaskExecutionState(final TaskExecutionState taskExecutionState) {

		boolean updateSuccess = executionGraph.updateState(taskExecutionState);

		if (updateSuccess) {
			final ExecutionVertexID executionVertexId = getExecutionVertexIdOrThrow(taskExecutionState.getID());
			if (isNotifiable(taskExecutionState)) {
				updateTaskExecutionStateInternal(executionVertexId, taskExecutionState);
			}
			return true;
		} else {
			return false;
		}
	}

	private boolean isNotifiable(final TaskExecutionState taskExecutionState) {
		final ExecutionState reportedExecutionState = taskExecutionState.getExecutionState();
		final ExecutionVertex executionVertex = getExecutionVertex(getExecutionVertexIdOrThrow(taskExecutionState.getID()));
		final ExecutionState effectiveExecutionState = executionVertex.getExecutionState();

		return (reportedExecutionState == FINISHED || reportedExecutionState == FAILED)
			&& effectiveExecutionState == reportedExecutionState;
	}
```

Reply:
We also potentially deserialize and serialize a `SeralizedThrowable`.

Reply:
Good idea. Will take this change.
Need to mention that we need to `getExecutionVertexIdOrThrow` before invoking `executionGraph.updateState` in case its a terminal state that deregisters the execution from execution graph.

Reply:
> Need to mention that we need to getExecutionVertexIdOrThrow before invoking 
> executionGraph.updateState in case its a terminal state that deregisters the execution 
> from execution graph.

ok didn't consider that

Reply:
updated.